### ISSUE TITLE
Writing to redis cluster is too slow(1million/1minute)

### ISSUE THREAD

**Issue by Eamon6221** [Grokzen/redis-py-cluster#371]
it's too long that written 1 billion data to my redis cluster by using this redis-py-cluster(the speed is about 1million/1minute).I wonder if this is normal？Thanks!

**Comment by Grokzen** [comment]
@Eamon6221 I dont know what your situation/setup/config is, and i have no reference point if this is fast or slow by comparing to something else. In general there is always an overhead when running a cluster client. Just the fact that you need to crc16 every key makes it slower. There is a benchmark script in the repo that you can run that should give you some kind of indication about your clusters raw performance.

A couple of optimizing tips. I assume that you do simple `SET ...` operations if you are bulking that many keys there will be a slowdown as there is a large roundtrip cost when doing a lot of keys at the same time. Two optimizations here is to bulk keys into a pipeline to send even more keys at once to each node in the cluster. The other optimization is to split your dataset into smaller chunks and run parralell clients that inserts keys so you can run it maybe x10 faster by running x10 clients doing their own work in parallel.

On a side note, if i remember the benchmarks correct, i did run maybe 20-25k SET/GET operations per second on my laptop against a cluster that was running locally. So move that to the cloud with more roundtrip and overhead etc would make 1 million operations per minute per client in line with what the performance should be around.

If you have any practical optimizations in the code that might improve any performance please open up a new issue and submit a PR to get the contribution included in the future.

**Comment by Eamon6221** [comment]
from rediscluster import RedisCluster
rc = RedisCluster(startup_nodes=startup_nodes, decode_responses=True)
pipe = rc.pipeline()
count=0
for result in hive_result:
       pipe.set('xxx','xxx',ex=xxx)
       if count%1000000==0:
             res = pipe.execute()
res = pipe.execute()
===============================================
Thank you for your reply! I used this way to write to my redis cluster.Is it correct? Thanks!

**Comment by Grokzen** [comment]
@Eamon6221 That will work, but i think that will have an insane bad performance writing 1 million keys per pipeline, try maybe 10k or 100k keys instead. But you are not going to find a much faster way to insert the keys really, unless you run it with some other compiled language or you do some pre processing of the keys so that the pipeline code don't have to do so much work, but i think you are limited to this unless you split this into multiple parallel executions.

The only other way to get any speedup in this scenario is if you pre-calculate all keys in one pass and then do the split of each key:value pair into the correct node they need to be sent to. Then you can go back to using the normal client Redis and just do a pipeline there and send all keys to each node. The downside is that you need to do a lot of the work that pipelines here gives you already, and you probably wont have any speedup anyway since you still need to do the key calculation for each given cluster anyway. One upside tho is that if you are rerunning this step multiple times, precalculating and saving the crc16 keys in one go will speed up your future executions, but note that the CRC16 calculations might differ if you go to another cluster with the data and not run on the same cluster setup and slot distribution.

**Comment by Grokzen** [comment]
Also i will reitterate that when using a redis cluster for the performance parts, you only get that from running at least 3 clients at the same time in my experience. The redis-server itself has no problems handling the load from it's side, the problem is the performance in the clients, and from experience and from the benchmarks, you need about 3 parallel clients to max out before the loss on each client will overtake the gains from using a clustered redis.

**Comment by Eamon6221** [comment]
Thanks for reply!The speed is about 12 millions/ 1 minute that my colleague written the same data into this redis cluster(It have 30 master nodes and 60 slave nodes) with Java. In fact,I had used  parallel clients like your benchmark script(I started 10 parallel clients). split the data into 10 parallel  client to write into redis cluster,but the performance just a little good. Before consulting you, I really want to use  the way of pre-calculate all keys to write into redis cluster,but I have no idea about how to do it detailly.☹

**Comment by Grokzen** [comment]
@Eamon6221 Look at the crc16 module https://github.com/Grokzen/redis-py-cluster/blob/master/rediscluster/crc.py thta is the code that does the key to slot calculation. Or you can just use this method https://github.com/Grokzen/redis-py-cluster/blob/master/rediscluster/nodemanager.py#L50 and pass in your key and it will return back what slot that key should belong to.

Well if you compare to java, python will always be slower compared to many other languages. It is not about the redis-server or the implementation logic in the client, but the slowness of the language itself that makes java work x12 faster then the same code in java basically. Maybe you are hitting 100% cpu or something when you push your x10 clients at the same time, and then you would not get any performance boost at all in that instance.

If you can use java for this task to import that many keys and the java solution is x12 times faster, then just go with java to do the import and save a ton of time. I know i would do it. There is no other real performance tricks deeper in that can be done really.

**Comment by Eamon6221** [comment]
OK. Thank you very much!🙂
